rero_ils/modules/notifications/dispatcher.py

```diff
@@ -270,6 +278,49 @@ def _process_notification(cls, notification, resend, aggregated):
         aggregation['documents'].append(documents_data)
         aggregation['notifications'].append(notification)
 
+    @staticmethod
+    def _can_cancel_notification(data, item):
+        """Check if a notification must be be canceled.
+
+        As notification process life cycle could work asynchronously, in some
+        case, when we process a notification, it's not yet required to
+        send it. By example, an AVAILABLE notification must not be send if the
+        related loan item is already in ON_LOAN state.
+
+        :param data: the notification data to check.
+        :param item: the notification related item
+        :return True if the notification can be canceled, False otheriwse.
+        """
+        cancel_notification = False
+
+        n_type = data['notification_type']
+        loan = data['loan']
+
+        # a) we don't need to check INTERNAL_NOTIFICATION. These notifications
+        #    are sent to library and are synchronously dispatched.
+        # b.1) Special case for RECALL notification: If the related loan isn't
+        #      yet ON_LOAN state, we doesn't need to process it.
+        # b.2) For OVERDUE and DUE_SOON notification, check that none other
+        #      corresponding notification already sent.
+        # b.3) Otherwise, check if the notification type is available into
+        #      notification candidates related to the loan. If not, the
+        #      notification can be canceled.
```

Comment:
```suggestion
        # b.3) Otherwise, check if the notification type is available into
        #      candidate notifications related to the loan. If not, the
        #      notification can be canceled.
```